src/lib/operators/table_scan/column_between_table_scan_impl.cpp

```diff
@@ -54,53 +56,70 @@ void ColumnBetweenTableScanImpl::_scan_generic_segment(const BaseSegment& segmen
+        return between_comparator_function(position.value(), typed_left_value, typed_right_value);
+      };
+      _scan_with_iterators<true>(between_comparator, it, end, chunk_id, matches);
+    });
   });
 }
 
 void ColumnBetweenTableScanImpl::_scan_dictionary_segment(const BaseDictionarySegment& segment, const ChunkID chunk_id,
                                                           PosList& matches,
                                                           const std::shared_ptr<const PosList>& position_filter) const {
-  auto left_value_id = segment.lower_bound(_left_value);
-  auto right_value_id = segment.upper_bound(_right_value);
+  ValueID lower_bound_value_id;
+  if (is_lower_inclusive_between(_predicate_condition)) {
+    lower_bound_value_id = segment.lower_bound(_left_value);
+  } else {
+    lower_bound_value_id = segment.upper_bound(_left_value);
+  }
 
-  if (right_value_id == INVALID_VALUE_ID) {
-    // lower/upper_bound returns INVALID_VALUE_ID for NULL, while the dictionary uses unique_values_count (#1283).
-    right_value_id = static_cast<ValueID>(segment.unique_values_count());
+  ValueID upper_bound_value_id;
+  if (is_upper_inclusive_between(_predicate_condition)) {
+    upper_bound_value_id = segment.upper_bound(_right_value);
+  } else {
+    upper_bound_value_id = segment.lower_bound(_right_value);
   }
 
-  auto column_iterable = create_iterable_from_attribute_vector(segment);
+  auto attribute_vector_iterable = create_iterable_from_attribute_vector(segment);
 
+  /**
+   * Early out: All entries (except NULLs) match
+   */
   // NOLINTNEXTLINE - cpplint is drunk
-  if (left_value_id == ValueID{0} && right_value_id == static_cast<ValueID>(segment.unique_values_count())) {
-    // all values match
-    column_iterable.with_iterators(position_filter, [&](auto left_it, auto left_end) {
+  if (lower_bound_value_id == ValueID{0} && upper_bound_value_id == INVALID_VALUE_ID) {
+    attribute_vector_iterable.with_iterators(position_filter, [&](auto left_it, auto left_end) {
       static const auto always_true = [](const auto&) { return true; };
       _scan_with_iterators<true>(always_true, left_it, left_end, chunk_id, matches);
     });
 
     return;
   }
 
-  if (left_value_id >= static_cast<ValueID>(segment.unique_values_count()) || left_value_id == right_value_id) {
-    // no values match
+  /**
+   * Early out: No entries match
+   */
+  if (lower_bound_value_id == INVALID_VALUE_ID || lower_bound_value_id >= upper_bound_value_id) {
     return;
   }
 
-  const auto value_id_diff = right_value_id - left_value_id;
+  /**
+   * No early out possible: Actually scan the attribute vector
+   */
 
-  const auto comparator = [left_value_id, value_id_diff](const auto& position) {
-    // Using < here because the right value id is the upper_bound. Also, because the value ids are integers, we can do
-    // a little hack here: (x >= a && x < b) === ((x - a) < (b - a)); cf. https://stackoverflow.com/a/17095534/2204581
```

Comment:
Any reason you removed this? This makes the difference between 21.255 iter/s and 29.1051 iter/s. Admittedly, the comment should have explained that this was done for performance reasons.

Reply:
@mrzzzrm

Reply:
Distant memories.... I think there was a reason, but it may well be that it became obsolete with the adjustment of upper_bound_value_id:

```c++
  if (upper_bound_value_id == INVALID_VALUE_ID) {
    upper_bound_value_id = segment.unique_values_count();
  }
```

Feel free to revert, if it doesn't trigger any tests it should be fine.